src/main/java/com/hms_networks/sc/cumulocity/api/CConnectorMqttMgr.java

```diff
@@ -77,15 +79,18 @@ public class CConnectorMqttMgr extends ConstrainedMqttManager {
    */
   private static final boolean MQTT_WAIT_FOR_WAN_IP = true;
 
-  /** MQTT loop attempt count. Used for retry backoff in {@link #runOnMqttLoop(int)} */
-  private int mqttLoopAttemptCount = 0;
-
-  /** MQTT loop limit, before triggering a reprovision event */
-  private int mqttLoopAttemptLimit = 15;
+  /**
+   * The maximum number of times to retry sending a message to Cumulocity before giving up and
+   * discarding the message (with warning).
+   */
+  private static final int PENDING_RETRY_MESSAGE_MAX_RETRY_COUNT = 16;
 
   /** List of child devices which have been registered to Cumulocity. */
   private final List registeredChildDevices = new ArrayList();
 
+  /** Stack of {@link CConnectorRetryMessage}s which have been queued for retry. */
+  private final Stack pendingRetryMessages = new Stack();
```

Comment:
Stack is LIFO. Perhaps FIFO is better?  Not sure that this matters.

Reply:
The goal with LIFO was to ensure that once the connection is re-established, the most recent data is more readily available.